File: app/models/pm/reports.py
from __future__ import annotations

from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, Field


# ════════════════════════════════════════════════════════════════════
# 수기 항목 (Manual Items for Weekly Reports)
# ════════════════════════════════════════════════════════════════════

class ManualItemCreate(BaseModel):
    section: str  # MAIN_AGENDA | ISSUE_RISK | DECISION_REQUIRED
    title: str
    owner: Optional[str] = None
    linked_sr_id: Optional[str] = None
    linked_issue_id: Optional[str] = None
    include_in_report: bool = True
    sort_order: int = 10
    # MAIN_AGENDA 전용
    category: Optional[str] = None
    content: Optional[str] = None
    agenda_status: Optional[str] = None
    # ISSUE_RISK 전용
    item_type: Optional[str] = None
    impact: Optional[str] = None
    action_plan: Optional[str] = None
    # DECISION_REQUIRED 전용
    background: Optional[str] = None
    options: Optional[str] = None
    requested_decision: Optional[str] = None
    desired_date: Optional[datetime] = None


class ManualItemPatch(BaseModel):
    title: Optional[str] = None
    owner: Optional[str] = None
    include_in_report: Optional[bool] = None
    sort_order: Optional[int] = None
    category: Optional[str] = None
    content: Optional[str] = None
    agenda_status: Optional[str] = None
    item_type: Optional[str] = None
    impact: Optional[str] = None
    action_plan: Optional[str] = None
    background: Optional[str] = None
    options: Optional[str] = None
    requested_decision: Optional[str] = None
    desired_date: Optional[datetime] = None


class ManualItemOut(BaseModel):
    id: str
    section: str
    title: str
    owner: Optional[str] = None
    linked_sr_id: Optional[str] = None
    linked_issue_id: Optional[str] = None
    include_in_report: bool = True
    sort_order: int = 0
    category: Optional[str] = None
    content: Optional[str] = None
    agenda_status: Optional[str] = None
    item_type: Optional[str] = None
    impact: Optional[str] = None
    action_plan: Optional[str] = None
    background: Optional[str] = None
    options: Optional[str] = None
    requested_decision: Optional[str] = None
    desired_date: Optional[datetime] = None
    created_by: str
    created_at: datetime
    updated_by: Optional[str] = None
    updated_at: datetime


class WeeklyReportStatusPatch(BaseModel):
    status: str  # DRAFT | REVIEWING | CONFIRMED


class SrItem(BaseModel):
    sr_no: str
    title: str
    status: str
    status_label: str
    request_type: str
    request_type_label: str
    requester_name: str
    requester_department: str
    assignee_name: Optional[str] = None
    is_urgent: bool = False
    desired_due_date: Optional[datetime] = None
    created_at: datetime


class SrSummary(BaseModel):
    new_this_week: List[SrItem] = Field(default_factory=list)      # 이번 주 신규 접수
    completed_this_week: List[SrItem] = Field(default_factory=list) # 이번 주 완료
    pending_items: List[SrItem] = Field(default_factory=list)       # 미접수 (SUBMITTED)
    open_items: List[SrItem] = Field(default_factory=list)          # 처리 중 전체 (비완료·비취소)
    by_status: dict = Field(default_factory=dict)                   # 상태별 카운트
    total_open: int = 0
    total_new: int = 0
    total_completed: int = 0


class ReportStats(BaseModel):
    total: int = 0
    completed: int = 0
    in_progress: int = 0
    delayed: int = 0
    todo: int = 0
    completion_rate: int = 0  # 정수 %


class WorkItem(BaseModel):
    issue_id: str
    issue_number: int
    title: str
    type: str
    project_id: str
    project_name: str
    org_name: Optional[str] = None
    assignee_id: Optional[str] = None
    assignee_name: Optional[str] = None
    status: str
    priority: str
    epic_title: Optional[str] = None
    sprint_name: Optional[str] = None
    start_date: Optional[datetime] = None
    due_date: Optional[datetime] = None
    is_delayed: bool = False
    story_points: Optional[int] = None


class ProjectBreakdown(BaseModel):
    project_id: str
    project_name: str
    org_name: Optional[str] = None
    stats: ReportStats = Field(default_factory=ReportStats)
    completed: List[WorkItem] = Field(default_factory=list)
    in_progress: List[WorkItem] = Field(default_factory=list)
    delayed: List[WorkItem] = Field(default_factory=list)
    upcoming: List[WorkItem] = Field(default_factory=list)


class PersonBreakdown(BaseModel):
    user_id: str
    user_name: str
    stats: ReportStats = Field(default_factory=ReportStats)
    completed: List[WorkItem] = Field(default_factory=list)
    in_progress: List[WorkItem] = Field(default_factory=list)
    delayed: List[WorkItem] = Field(default_factory=list)
    upcoming: List[WorkItem] = Field(default_factory=list)


# ════════════════════════════════════════════════════════════════════
# 주간 보고
# ════════════════════════════════════════════════════════════════════

class WeeklyReportCreate(BaseModel):
    report_year: int
    report_week: int
    start_date: datetime
    end_date: datetime
    title: str
    department: Optional[str] = None
    admin_comment: Optional[str] = None


class WeeklyReportPatch(BaseModel):
    title: Optional[str] = None
    department: Optional[str] = None
    admin_comment: Optional[str] = None


class WeeklyReportOut(BaseModel):
    id: str
    report_year: int
    report_week: int
    start_date: datetime
    end_date: datetime
    title: str
    department: Optional[str] = None
    status: str = "DRAFT"
    by_project: List[ProjectBreakdown] = Field(default_factory=list)
    by_person: List[PersonBreakdown] = Field(default_factory=list)
    all_items: List[WorkItem] = Field(default_factory=list)
    upcoming_items: List[WorkItem] = Field(default_factory=list)
    stats: ReportStats = Field(default_factory=ReportStats)
    manual_items: List[ManualItemOut] = Field(default_factory=list)
    sr_summary: Optional[SrSummary] = None
    admin_comment: Optional[str] = None
    created_by: str
    created_by_name: Optional[str] = None
    created_at: datetime
    updated_by: Optional[str] = None
    updated_by_name: Optional[str] = None
    updated_at: datetime
    confirmed_by: Optional[str] = None
    confirmed_at: Optional[datetime] = None


# ════════════════════════════════════════════════════════════════════
# 월간 보고
# ════════════════════════════════════════════════════════════════════

class MonthlyReportCreate(BaseModel):
    report_year: int
    report_month: int
    title: str
    department: Optional[str] = None
    admin_comment: Optional[str] = None


class MonthlyReportPatch(BaseModel):
    title: Optional[str] = None
    department: Optional[str] = None
    admin_comment: Optional[str] = None


class MonthlyReportOut(BaseModel):
    id: str
    report_year: int
    report_month: int
    title: str
    department: Optional[str] = None
    by_project: List[ProjectBreakdown] = Field(default_factory=list)
    by_person: List[PersonBreakdown] = Field(default_factory=list)
    all_items: List[WorkItem] = Field(default_factory=list)
    upcoming_items: List[WorkItem] = Field(default_factory=list)
    stats: ReportStats = Field(default_factory=ReportStats)
    admin_comment: Optional[str] = None
    created_by: str
    created_by_name: Optional[str] = None
    created_at: datetime
    updated_by: Optional[str] = None
    updated_by_name: Optional[str] = None
    updated_at: datetime

File: frontend/optool/src/services/pm/reports.ts
import { api } from 'boot/axios'

export interface ReportStats {
  total: number
  completed: number
  inProgress: number
  delayed: number
  todo: number
  completionRate: number
}

export interface WorkItem {
  issueId: string
  issueNumber: number
  title: string
  type: string
  projectId: string
  projectName: string
  orgName: string | null
  assigneeId: string | null
  assigneeName: string | null
  status: string
  priority: string
  epicTitle: string | null
  sprintName: string | null
  startDate: string | null
  dueDate: string | null
  isDelayed: boolean
  storyPoints: number | null
}

export interface ProjectBreakdown {
  projectId: string
  projectName: string
  orgName: string | null
  stats: ReportStats
  completed: WorkItem[]
  inProgress: WorkItem[]
  delayed: WorkItem[]
  upcoming: WorkItem[]
}

export interface PersonBreakdown {
  userId: string
  userName: string
  stats: ReportStats
  completed: WorkItem[]
  inProgress: WorkItem[]
  delayed: WorkItem[]
  upcoming: WorkItem[]
}

// ── 수기 항목 ────────────────────────────────────────────────────────
export type ReportStatus = 'DRAFT' | 'REVIEWING' | 'CONFIRMED'
export type ManualItemSection = 'MAIN_AGENDA' | 'ISSUE_RISK' | 'DECISION_REQUIRED' | 'ANNOUNCEMENT' | 'ATTENDANCE'

export interface ManualItem {
  id: string
  section: ManualItemSection
  title: string
  owner: string | null
  linkedSrId: string | null
  linkedIssueId: string | null
  includeInReport: boolean
  sortOrder: number
  // MAIN_AGENDA
  category: string | null
  content: string | null
  agendaStatus: string | null
  // ISSUE_RISK
  itemType: string | null
  impact: string | null
  actionPlan: string | null
  // DECISION_REQUIRED
  background: string | null
  options: string | null
  requestedDecision: string | null
  desiredDate: string | null
  createdBy: string
  createdAt: string
  updatedBy: string | null
  updatedAt: string
}

export interface ManualItemCreate {
  section: ManualItemSection
  title: string
  owner?: string | null
  include_in_report?: boolean
  sort_order?: number
  category?: string | null
  content?: string | null
  agenda_status?: string | null
  item_type?: string | null
  impact?: string | null
  action_plan?: string | null
  background?: string | null
  options?: string | null
  requested_decision?: string | null
  desired_date?: string | null
}

// ── SR 요약 ──────────────────────────────────────────────────────────
export interface SrItem {
  srNo: string
  title: string
  status: string
  statusLabel: string
  requestType: string
  requestTypeLabel: string
  requesterName: string
  requesterDepartment: string
  assigneeName: string | null
  isUrgent: boolean
  desiredDueDate: string | null
  createdAt: string
}

export interface SrSummary {
  newThisWeek: SrItem[]
  completedThisWeek: SrItem[]
  pendingItems: SrItem[]
  openItems: SrItem[]
  byStatus: Record<string, number>
  totalOpen: number
  totalNew: number
  totalCompleted: number
}

// ── 주간 보고 ────────────────────────────────────────────────────────
export interface WeeklyReport {
  id: string
  reportYear: number
  reportWeek: number
  startDate: string
  endDate: string
  title: string
  department: string | null
  status: ReportStatus
  byProject: ProjectBreakdown[]
  byPerson: PersonBreakdown[]
  allItems: WorkItem[]
  upcomingItems: WorkItem[]
  stats: ReportStats
  manualItems: ManualItem[]
  srSummary: SrSummary | null
  adminComment: string | null
  createdBy: string
  createdByName: string | null
  createdAt: string
  updatedBy: string | null
  updatedByName: string | null
  updatedAt: string
  confirmedBy: string | null
  confirmedAt: string | null
}

export interface WeeklyReportCreate {
  report_year: number
  report_week: number
  start_date: string
  end_date: string
  title: string
  department?: string
  admin_comment?: string
}

export interface WeeklyReportPatch {
  title?: string
  department?: string
  admin_comment?: string
}

export async function listWeeklyReports(params: { year?: number; week?: number; status?: string } = {}) {
  const { data } = await api.get<WeeklyReport[]>('/pm/weekly-reports', { params })
  return data
}

export async function getWeeklyReport(id: string) {
  const { data } = await api.get<WeeklyReport>(`/pm/weekly-reports/${id}`)
  return data
}

export async function createWeeklyReport(payload: WeeklyReportCreate) {
  const { data } = await api.post<WeeklyReport>('/pm/weekly-reports', payload)
  return data
}

export async function updateWeeklyReport(id: string, payload: WeeklyReportPatch) {
  const { data } = await api.patch<WeeklyReport>(`/pm/weekly-reports/${id}`, payload)
  return data
}

export async function refreshWeeklyReport(id: string) {
  const { data } = await api.post<WeeklyReport>(`/pm/weekly-reports/${id}/refresh`)
  return data
}

export async function deleteWeeklyReport(id: string) {
  await api.delete(`/pm/weekly-reports/${id}`)
}

export async function changeWeeklyReportStatus(id: string, status: ReportStatus) {
  const { data } = await api.patch<WeeklyReport>(`/pm/weekly-reports/${id}/status`, { status })
  return data
}

export async function addManualItem(id: string, payload: ManualItemCreate) {
  const { data } = await api.post<WeeklyReport>(`/pm/weekly-reports/${id}/items`, payload)
  return data
}

export async function updateManualItem(id: string, itemId: string, payload: Partial<ManualItemCreate>) {
  const { data } = await api.patch<WeeklyReport>(`/pm/weekly-reports/${id}/items/${itemId}`, payload)
  return data
}

export async function deleteManualItem(id: string, itemId: string) {
  const { data } = await api.delete<WeeklyReport>(`/pm/weekly-reports/${id}/items/${itemId}`)
  return data
}

export async function previewWeeklyReport(id: string): Promise<string> {
  const { data } = await api.get<{ text: string }>(`/pm/weekly-reports/${id}/preview`)
  return data.text
}

export function weeklyExportListUrl(params: { year?: number; week?: number } = {}) {
  const qs = new URLSearchParams()
  if (params.year) qs.set('year', String(params.year))
  if (params.week) qs.set('week', String(params.week))
  return `/api/pm/weekly-reports/export/list?${qs}`
}

export function weeklyExportDetailUrl(id: string) {
  return `/api/pm/weekly-reports/${id}/export`
}

// ── 월간 보고 ────────────────────────────────────────────────────────
export interface MonthlyReport {
  id: string
  reportYear: number
  reportMonth: number
  title: string
  department: string | null
  byProject: ProjectBreakdown[]
  byPerson: PersonBreakdown[]
  allItems: WorkItem[]
  upcomingItems: WorkItem[]
  stats: ReportStats
  adminComment: string | null
  createdBy: string
  createdByName: string | null
  createdAt: string
  updatedBy: string | null
  updatedByName: string | null
  updatedAt: string
}

export interface MonthlyReportCreate {
  report_year: number
  report_month: number
  title: string
  department?: string
  admin_comment?: string
}

export interface MonthlyReportPatch {
  title?: string
  department?: string
  admin_comment?: string
}

export async function listMonthlyReports(params: { year?: number; month?: number; status?: string } = {}) {
  const { data } = await api.get<MonthlyReport[]>('/pm/monthly-reports', { params })
  return data
}

export async function getMonthlyReport(id: string) {
  const { data } = await api.get<MonthlyReport>(`/pm/monthly-reports/${id}`)
  return data
}

export async function createMonthlyReport(payload: MonthlyReportCreate) {
  const { data } = await api.post<MonthlyReport>('/pm/monthly-reports', payload)
  return data
}

export async function updateMonthlyReport(id: string, payload: MonthlyReportPatch) {
  const { data } = await api.patch<MonthlyReport>(`/pm/monthly-reports/${id}`, payload)
  return data
}

export async function refreshMonthlyReport(id: string) {
  const { data } = await api.post<MonthlyReport>(`/pm/monthly-reports/${id}/refresh`)
  return data
}

export async function deleteMonthlyReport(id: string) {
  await api.delete(`/pm/monthly-reports/${id}`)
}

export function monthlyExportListUrl(params: { year?: number; month?: number } = {}) {
  const qs = new URLSearchParams()
  if (params.year)  qs.set('year',  String(params.year))
  if (params.month) qs.set('month', String(params.month))
  return `/api/pm/monthly-reports/export/list?${qs}`
}

export function monthlyExportDetailUrl(id: string) {
  return `/api/pm/monthly-reports/${id}/export`
}
